### ISSUE TITLE
prefer_const_constructors_in_immutables needs updates for 0.36.1

### ISSUE THREAD

**Issue by pq** [dart-lang/linter#1537]
With the `package:analyzer` v 0.36.1, we're seeing new failures in `prefer_const_constructors_in_immutables` tests.  Specifically, this is no longer linting:

```dart
@immutable
class H {
  final f;
  H(f) : f = f ?? f == null; // LINT
}
```

See: https://travis-ci.org/dart-lang/linter/jobs/523234741

/cc @bwilkerson @a14n

**Comment by pq** [comment]
Marking as a P1, since we'll want to fix this before publishing a new linter (#1535).

**Comment by pq** [comment]
After looking this over w/ @scheglov, it appears that the hack used in `_hasOnlyConstExpressionsInIntializerList` took advantage of accidental behavior in the analyzer and will no longer work --- the rub is that these programmatic changes to the AST are no longer being reflected in the element model.  We discussed a few options.  One involves changing `ConstantVerifier` to allow clients to state that they want an element to be treated as constant.  Maybe there are others?

@a14n: any thoughts?

To move ahead with the release, I'll update the tests to note that these cases are no longer being caught and link back here for updating once this issue is definitively addressed.

**Comment by a14n** [comment]
Note that the same kind of hack is used in `unnecessary_const` lint.

**Comment by a14n** [comment]
>Note that the same kind of hack is used in `unnecessary_const` lint.

And `prefer_const_literals_to_create_immutables` and `prefer_const_constructors` and `prefer_const_constructors_in_immutables`.

**Comment by pq** [comment]
Ah.  Thanks for that added context @a14n.  Without looking, it's interesting that we're not seeing failures there too.  Maybe we're not testing those paths?  Or maybe this functionality is still (accidentally) working in those cases?

Either way, it seems like this is significant enough that we should figure out how to get this information in a more reliable way.  @bwilkerson mentioned that `AstNode` used to have this info.  Maybe we could add that back?  Or provide some other means?

@scheglov: any thoughts?

**Comment by scheglov** [comment]
I will rewrite it without looking into the enclosing `ConstructorElement`.

**Comment by scheglov** [comment]
https://dart-review.googlesource.com/c/sdk/+/100277

I have not tried if this fixed linter.

**Comment by pq** [comment]
Thank you @scheglov!

With that fix, the existing tests pass and this now gets flagged:

```dart
@immutable
class D {
  final _a;
  D.c4(a) : _a = '${a ? a : ''}'; // OK --- but should actually be a LINT?
}
```

Do we agree that linting here is the correct behavior?

**Comment by bwilkerson** [comment]
I thought that the conditional expression wasn't added as a constant until constant-update-2018 (which hasn't landed yet). But if they were allowed before that, then the expression ought to be valid.

**Comment by scheglov** [comment]
I see it in the [spec](https://spec.dart.dev/DartLangSpecDraft.pdf).
![image](https://user-images.githubusercontent.com/384794/56623089-a1540680-65e7-11e9-9e4d-27f0e9a03d9c.png)

And I don't see it in [constants update 2018](https://github.com/dart-lang/language/blob/master/accepted/future-releases/constant-update-2018/feature-specification.md).

**Comment by scheglov** [comment]
Unfortunately the CL fails tests :-(

The reason is that when we verify constants,  we use constructor initializers resynthesized from summary, and so no attached to `ConstructorDeclaration`. So, there is nothing to check for `constKeyword` presence.

**Comment by eernstg** [comment]
@scheglov wrote:
> And I don't see it in [constants update 2018](https://github.com/dart-lang/language/blob/master/accepted/future-releases/constant-update-2018/feature-specification.md).

It is actually mentioned in said feature spec:

> Finally, the conditional `?`/`:` operator only evaluates one of
> its branches, depending on whether the condition expression
> evaluates to true or false. The other branch must also be a
> potenatially constant expression.

These changes were introduced into the spec with the SDK commit 0808d29b4e33c303bf40524407a6e331c7a439f8 on Sep 26 2018, in the expectation that the 'constant update 2018' would be completed in 2018.

The bullet point on string interpolation was also changed in 0808d29b4e33c303bf40524407a6e331c7a439f8, for the same reason.

So, if this is helpful, those changes could certainly be considered to be a part of constant-update-2018, and it would be appropriate to flag usages of them as errors when that feature has not been enabled.

**Comment by pq** [comment]
> Unfortunately the CL fails tests 

😢 

@scheglov: could we consider a param to `ConstantVerifier` to get this behavior just when desired?

As for the spec and `constant-update-2018` thanks a million @eernstg for the clarifications!

**Comment by pq** [comment]
@a14n: #1551 should address this issue.  Are there other places that would benefit from the same treatment?

**Comment by pq** [comment]
Closing for now.  @a14n: please open follow-ups if needed?  Thanks!

(And cheers @scheglov for the analyzer fix!  🍻)

**Comment by a14n** [comment]
Follow-ups:

- #1554 
- #1555 
- #1556
